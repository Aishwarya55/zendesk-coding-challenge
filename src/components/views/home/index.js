import React, { Component } from 'react'
import Subscription from "./../subscription"
import SubscriptionSubmission from "./../subscription-submission"
import {
    Switch,
    Route,
    Redirect
} from 'react-router-dom'
import "./home.css"

class Home extends Component {


    render() {
        return <div className="homeContainer"><Switch>
            <Route exact path="/home">
                <Redirect to="/home/subscription" />
            </Route>
            <Route exact path="/home/subscription" component={Subscription} />
            <Route exact path="/home/submission" component={SubscriptionSubmission}>
                
        </Route>

        </Switch>
    
        </div>
    }
}

export default Home