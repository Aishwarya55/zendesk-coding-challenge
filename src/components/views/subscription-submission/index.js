import React, { Component } from 'react'
import { getPrice, getCurrentSubscription } from "../../../actions/index"
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import "./subscription-submission.css"
import { withRouter } from "react-router-dom"
export class SubscriptionSubmission extends Component {

    goBack() {
        this.props.history.push("/home/subscription")
    }

    componentDidMount(){
        //Redirect page when accessed directly without updating subscription
        if(this.props.currentSubscription.cost.length === 0){
            this.props.history.push("/home/subscription")
        }
    }

    render() {
        return <>

            <div className="subscriptionsubContainer">
                <span className="back-btn" onClick={() => { this.goBack() }}>&lt;&nbsp;&nbsp;Back to Subscription</span>
                <div className="subscriptionSubContainer prev">
                    <h4>Previous Subscription</h4>
                    <div>
                        <span>Plan</span>
                        <span>{this.props.previousSubscription.name}</span>
                    </div>
                    <div>
                        <span>Seats</span>
                        <span>{this.props.previousSubscription.seats}</span>
                    </div>
                    <div>
                        <span>Price</span>
                        <span>${this.props.previousSubscription.cost}</span>
                    </div>

                </div>

                <div className="subscriptionSubContainer curr">
                    <h4>Current Subscription</h4>
                    <div>
                        <span>Plan</span>
                        <span className="currValue">{this.props.currentSubscription.name}</span>
                    </div>
                    <div>
                        <span>Seats</span>
                        <span className="currValue">{this.props.currentSubscription.seats}</span>
                    </div>
                    <div>
                        <span>Price</span>
                        <span className="currValue">${this.props.currentSubscription.cost}</span>
                    </div>

                </div>
            </div>
        </>
    }
}

const mapStateToProps = (state) => {
    return {

        previousSubscription: state.subscriptionReducer.previousSubscription,
        currentSubscription: state.subscriptionReducer.currentSubscription
    }
}

const mapDispatchToProps = dispatch => bindActionCreators({
    getPrice,
    getCurrentSubscription
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(SubscriptionSubmission));