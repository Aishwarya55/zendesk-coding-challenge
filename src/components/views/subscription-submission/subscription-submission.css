.subscriptionSubContainer div {
  display: flex;
  margin-bottom: 16px;
}

.subscriptionSubContainer span {
  flex: 1;
}

.subscriptionsubContainer {
  display: flex;
  flex-direction: column;
  width: fit-content;
  min-width: 400px;
  padding: 32px;
  position: absolute;
  top: 45%;
  left: 50%;
  -webkit-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  background: #fff;
}

.subscriptionSubContainer {
  margin: 16px 0;
  min-width: 250px;
  padding: 32px;

  text-align: center;
  background: #fff;
  box-shadow: 0 1px 10px 1px rgb(0, 0, 0, 0.09);
}

.curr {
  background-color: rgba(23, 162, 184, 0.4);
}

.currValue {
  color: #0e5e6b;
  font-weight: 700;
}

.back-btn {
  font-size: 13px;
  cursor: pointer;
  color: #17a2b8;
  font-weight: 500;
  padding: 8px 0;
}

.back-btn:hover {
  color: #138496;
}
