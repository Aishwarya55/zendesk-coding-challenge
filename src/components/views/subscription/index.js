import React, { Component } from 'react'
import { getPrice, getCurrentSubscription, updateSubscription } from "../../../actions/index"
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from "react-router-dom"
import "./subscription.css"
import Loader from "./../../loader/index"
export class Subscription extends Component {



    state = {
        subscriptionData: {
            plan: "basic",
            seats: 0
        },
        isEnabled: false
    }

    componentDidMount() {

        this.props.getCurrentSubscription()
    }


    changeSubscription(type, event, subscriptionData) {



        if (type === "seats") {
            let num = event.target.value.match(/^\d+$/);
            if (num === null) {

                event.target.value = "";
            }

            if (event.target.value === "0") {
                event.target.value = "1";
            }
        }


        let newSubcriptionData = {
            ...subscriptionData,
            [type]: event.target.value
        }
        let fieldDirty = false


        Object.keys(newSubcriptionData).forEach(key => {

            if (newSubcriptionData[key] !== this.props.currentSubscription[key])
                fieldDirty = true

        })

        this.setState({
            ...this.state,
            subscriptionData: {
                ...subscriptionData,
                [type]: event.target.value
            },
            isEnabled: fieldDirty
        })


        if (event.target.value.length !== 0)
            this.props.getPrice({
                ...newSubcriptionData
            })

    }

    validate(e, subscriptionData) {
        if (e.target.value.trim().length === 0) {
            e.target.value = this.props.currentSubscription.seats
            this.changeSubscription("seats", e, subscriptionData)
        }
    }

    componentDidUpdate(prevProp, prevState) {



    }

    updateNewSubscription() {
        this.props.updateSubscription({
            ...this.props.currentSubscription,
            ...this.state.subscriptionData
        }).then(success => {
            this.props.history.push("/home/submission")
        })
    }




    render() {

        const subscriptionData = !this.state.isEnabled ? this.props.currentSubscription : this.state.subscriptionData

        const costPreview = !this.state.isEnabled ? this.props.currentSubscription : this.props.costPreview
        return <>
            <div class="card">
                <h2>
                    Subscription
        </h2>
                <div className="outerContainer">

                    <div className="subscriptionContainer">
                        <div className="formElement">
                            <label htmlFor="plan" className="label">
                                Plan
            </label>
                            <select name="plan" className="planSelect formComp" value={subscriptionData.plan} onChange={(e) => { this.changeSubscription("plan", e, subscriptionData) }}>
                                <option value="basic">Basic</option>
                                <option value="good">Good</option>
                                <option value="better">Better</option>
                                <option value="best">Best</option>
                            </select>

                        </div>


                        <div className="formElement">
                            <label htmlFor="seat" className="label">
                                Seats
            </label>
                            <input className="seatInput formComp" type="number" name="seat" min="1" value={subscriptionData.seats} onChange={(e) => { this.changeSubscription("seats", e, subscriptionData) }} onBlur={(e) => { this.validate(e, subscriptionData) }} />

                        </div>

                        <div className="formElement">

                            <div className="formComp price">
                                Price
            </div>
                            <div className="formComp costEl">
                                {!this.props.loading && `$${costPreview.cost}`}
                                <Loader showLoader={this.props.loading} />
                            </div>
                        </div>



                    </div>
                    <div className="buttonContainer">
                        <button disabled={!this.state.isEnabled || this.props.loading} onClick={() => this.updateNewSubscription()} >Update Subscription</button>
                    </div>
                </div>
            </div>
        </>
    }

}

const mapStateToProps = (state) => {
    return {
        costPreview: state.subscriptionReducer.costPreview,
        currentSubscription: state.subscriptionReducer.currentSubscription,
        loading: state.subscriptionReducer.loading
    }
}

const mapDispatchToProps = dispatch => bindActionCreators({
    getPrice,
    getCurrentSubscription,
    updateSubscription
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Subscription));