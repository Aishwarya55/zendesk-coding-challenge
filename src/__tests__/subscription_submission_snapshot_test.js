import React from "react"
import toJson from "enzyme-to-json"
import {shallow, configure} from "enzyme"
import Adapter from 'enzyme-adapter-react-16';
import {SubscriptionSubmission} from "../../src/components/views/subscription-submission"

configure({adapter: new Adapter()});
it('Renders correctly', () => {
let [getPrice, getCurrentSubscription] = new Array(2).fill(jest.fn());
const historyMock = { push: jest.fn() };
const props = {
    previousSubscription: {
        plan: "good",
        name: "Good",
        seats: "5",
        cost: "50"
    },
    currentSubscription: {
        plan: "best",
        name: "Best",
        seats: "5",
        cost: "5000"
    },
    history: historyMock
}
    const wrapper = shallow(
      <SubscriptionSubmission
       {...props}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
})