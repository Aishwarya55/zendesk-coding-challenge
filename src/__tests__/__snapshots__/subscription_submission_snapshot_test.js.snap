// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<Fragment>
  <div
    className="subscriptionsubContainer"
  >
    <span
      className="back-btn"
      onClick={[Function]}
    >
      &lt;  Back to Subscription
    </span>
    <div
      className="subscriptionSubContainer prev"
    >
      <h4>
        Previous Subscription
      </h4>
      <div>
        <span>
          Plan
        </span>
        <span>
          Good
        </span>
      </div>
      <div>
        <span>
          Seats
        </span>
        <span>
          5
        </span>
      </div>
      <div>
        <span>
          Price
        </span>
        <span>
          $
          50
        </span>
      </div>
    </div>
    <div
      className="subscriptionSubContainer curr"
    >
      <h4>
        Current Subscription
      </h4>
      <div>
        <span>
          Plan
        </span>
        <span
          className="currValue"
        >
          Best
        </span>
      </div>
      <div>
        <span>
          Seats
        </span>
        <span
          className="currValue"
        >
          5
        </span>
      </div>
      <div>
        <span>
          Price
        </span>
        <span
          className="currValue"
        >
          $
          5000
        </span>
      </div>
    </div>
  </div>
</Fragment>
`;
